File: WebClient SDK/src/default/views/viewsContextMenuExtension.ts
import { WorkplaceViewComponentExtension } from 'tessa/ui/views/extensions';
import { IWorkplaceViewComponent, IViewContextMenuContext } from 'tessa/ui/views';
import { MenuAction, SeparatorMenuAction, LoadingOverlay } from 'tessa/ui';
import { ViewReferenceMetadataSealed } from 'tessa/views/metadata';
import { getValueId, getFirstStringValueByPrefix } from 'tessa/views';
import { LocalizationManager } from 'tessa/localization';
import { openCard } from 'tessa/ui/uiHost';

export class ViewsContextMenuExtension extends WorkplaceViewComponentExtension {

  public getExtensionName(): string {
    return 'Tessa.Extensions.Default.Client.Views.ViewsContextMenuExtension';
  }

  public initialize(model: IWorkplaceViewComponent) {
    model.contextMenuGenerators.push(ctx => {
      ViewsContextMenuExtension.createRefreshMenuAction(ctx);
      ViewsContextMenuExtension.createFilterMenuAction(ctx);
      ViewsContextMenuExtension.createClearFilterMenuAction(ctx);
      ViewsContextMenuExtension.createOpenCardMenuAction(ctx);
    });
  }

  private static createRefreshMenuAction(ctx: IViewContextMenuContext) {
    const viewContext = ctx.viewContext;

    ctx.menuActions.push(new MenuAction(
      'Refresh',
      '$UI_Tiles_Refresh',
      'ta icon-thin-412',
      () => { viewContext.refreshView(); },
      null,
      !viewContext.canRefreshView()
    ));
  }

  private static createFilterMenuAction(ctx: IViewContextMenuContext) {
    const viewContext = ctx.viewContext;

    if (viewContext.parameters.metadata.every(p => p.hidden)) {
      return;
    }

    ctx.menuActions.push(new SeparatorMenuAction());
    ctx.menuActions.push(new MenuAction(
      'Filter',
      '$UI_Tiles_Filter',
      'ta icon-thin-100',
      () => { viewContext.filterView(); },
      null,
      !viewContext.canFilterView()
    ));
  }

  private static createClearFilterMenuAction(ctx: IViewContextMenuContext) {
    const viewContext = ctx.viewContext;

    if (viewContext.parameters.parameters.every(p => p.readOnly)) {
      return;
    }

    ctx.menuActions.push(new MenuAction(
      'ClearFilter',
      '$Views_Table_ClearFilter_ToolTip',
      'ta icon-thin-065',
      () => { viewContext.clearFilterView(); },
      null,
      !viewContext.canClearFilterView()
    ));
  }

  private static createOpenCardMenuAction(ctx: IViewContextMenuContext) {
    const viewContext = ctx.viewContext;

    if (viewContext.refSection) {
      return;
    }

    const view = viewContext.view;
    const selectedObject = viewContext.selectedRow;
    if (!selectedObject || !view) {
      return;
    }

    const metadata = view.metadata;
    if (!metadata) {
      return;
    }

    const cardRefs: ViewReferenceMetadataSealed[] = [];
    metadata.references.forEach(ref => {
      if (ref.isCard) {
        cardRefs.push(ref);
      }
    });
    if (cardRefs.length === 0) {
      return;
    }

    let separatorAddided = false;
    const addedIDs: Set<guid> = new Set();
    for (let cardRef of cardRefs) {
      const cardId = getValueId(selectedObject, cardRef.colPrefix!);
      if (!cardId) {
        return;
      }

      const displayValueObj = cardRef.displayValueColumn
        ? selectedObject.get(cardRef.displayValueColumn)
        : getFirstStringValueByPrefix(selectedObject, cardRef.colPrefix!, viewContext.columns);

      const displayValue = displayValueObj || '';

      if (!addedIDs.has(cardId)) {
        addedIDs.add(cardId);
        if (!separatorAddided) {
          ctx.menuActions.push(new SeparatorMenuAction());
          separatorAddided = true;
        }

        const uiContextExecutor = ctx.uiContextExecutor;
        ctx.menuActions.push(new MenuAction(
          `OpenCard_${cardId}`,
          `${LocalizationManager.instance.localize('$UI_Tiles_OpenCard')}: ${LocalizationManager.instance.localize(displayValue)}`,
          'ta icon-thin-013',
          () => {
            LoadingOverlay.instance.show(async (splashResolve) => {
              await uiContextExecutor(async context => {
                await openCard({
                  cardId,
                  displayValue,
                  context,
                  splashResolve
                });
              });
            });
          },
          null,
          false
        ));
      }
    }
  }

}

File: Source/Extensions/Tessa.Extensions.Default.Client/Views/ViewsContextMenuExtension.cs
using System;
using System.Collections.Generic;
using System.Diagnostics.Contracts;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Tessa.Localization;
using Tessa.Platform;
using Tessa.Platform.Collections;
using Tessa.Properties.Resharper;
using Tessa.UI;
using Tessa.UI.Menu;
using Tessa.UI.Views;
using Tessa.UI.Views.Content;
using Tessa.UI.Views.Workplaces.Tree;
using Tessa.Views;

namespace Tessa.Extensions.Default.Client.Views
{
    /// <summary>
    ///     Расширение демонстрирующее возможность добавления элементов контекстного меню
    /// </summary>
    /// <remarks>
    /// У расширения есть конфигуратор <see cref="ViewsContextMenuExtensionConfigurator"/>.
    /// </remarks>
    public sealed class ViewsContextMenuExtension : IWorkplaceViewComponentExtension
    {
        /// <summary>
        ///     The clear filter caption.
        /// </summary>
        private const string ClearFilterCaption = "$Views_Table_ClearFilter_ToolTip";

        /// <summary>
        ///     The clear filter icon name.
        /// </summary>
        private const string ClearFilterIconName = "Thin65";

        /// <summary>
        ///     The clear filter name.
        /// </summary>
        private const string ClearFilterName = "ClearFilter";

        /// <summary>
        ///     The filter icon name.
        /// </summary>
        private const string FilterIconName = "Thin100";

        /// <summary>
        ///     The filter name.
        /// </summary>
        private const string FilterName = "Filter";

        /// <summary>
        ///     The filter separator name.
        /// </summary>
        private const string FilterSeparatorName = "FilterSeparator";

        /// <summary>
        ///     The open card icon name.
        /// </summary>
        private const string OpenCardIconName = "Thin13";

        /// <summary>
        ///     The open card separator name.
        /// </summary>
        private const string OpenCardSeparatorName = "OpenCardSeparator";

        /// <summary>
        ///     The refresh icon name.
        /// </summary>
        private const string RefreshIconName = "Thin412";

        /// <summary>
        ///     The refresh name.
        /// </summary>
        private const string RefreshName = "Refresh";

        /// <summary>
        ///     The ui tiles filter.
        /// </summary>
        private const string UITilesFilter = "$UI_Tiles_Filter";

        /// <summary>
        ///     The ui tiles opencard.
        /// </summary>
        private const string UITilesOpenCard = "$UI_Tiles_OpenCard";

        /// <summary>
        ///     The ui tiles refresh.
        /// </summary>
        private const string UITilesRefresh = "$UI_Tiles_Refresh";

        /// <summary>
        ///     Gets or sets the host getter.
        /// </summary>
        private readonly Func<IUIHost> hostGetter;

        /// <summary>
        /// Initializes a new instance of the <see cref="ViewsContextMenuExtension"/> class.
        ///     Инициализирует новый экземпляр класса <see cref="T:System.Object"/>.
        /// </summary>
        /// <param name="hostGetter">
        /// Фабрика получения <see cref="IUIHost"/>
        /// </param>
        public ViewsContextMenuExtension([NotNull] Func<IUIHost> hostGetter)
        {
            Contract.Requires(hostGetter != null);
            this.hostGetter = hostGetter;
        }

        /// <summary>
        /// Вызывается при клонировании модели <paramref name="source"/> в контексте <paramref name="context"/>
        /// </summary>
        /// <param name="source">
        /// Исходная модель
        /// </param>
        /// <param name="cloned">
        /// Клонированная модель
        /// </param>
        /// <param name="context">
        /// Контекст клонирования
        /// </param>
        public void Clone(IWorkplaceViewComponent source, IWorkplaceViewComponent cloned, ICloneableContext context)
        {
        }

        /// <summary>
        /// Вызывается после создания модели <paramref name="model"/>
        /// </summary>
        /// <param name="model">
        /// Инициализируемая модель
        /// </param>
        public void Initialize(IWorkplaceViewComponent model)
        {
            model.ContextMenuGenerators.AddRange(
                GetRefreshMenuAction(),
                GetFilterMenuAction(),
                GetClearFilterMenuAction(),
                this.GetOpenCardMenuAction());
        }

        /// <summary>
        /// Вызывается после создания модели <paramref name="model"/> перед отображении в UI
        /// </summary>
        /// <param name="model">
        /// Модель
        /// </param>
        public void Initialized(IWorkplaceViewComponent model)
        {
        }

        /// <summary>
        ///     Возвращает действие создающее элемент меню
        /// </summary>
        /// <returns>
        ///     Действие создающее элемент меню
        /// </returns>
        private static Func<ViewContextMenuContext, ValueTask> GetClearFilterMenuAction()
        {
            return c =>
                {
                    if (c.ViewContext.Parameters.All(p => p.ReadOnly))
                    {
                        return new ValueTask();
                    }

                    c.MenuActions.Add(
                        new MenuAction(
                            ClearFilterName,
                            LocalizationManager.Localize(ClearFilterCaption),
                            c.MenuContext.Icons.Get(ClearFilterIconName),
                            new DelegateCommand(
                                async o => await c.ViewContext.ClearFilterAsync(c.ViewContext.Parameters),
                                o => c.ViewContext.CanClearFilter(c.ViewContext.Parameters))));

                    return new ValueTask();
                };
        }

        /// <summary>
        ///     Возвращает действие создающее элемент меню
        /// </summary>
        /// <returns>
        ///     Действие создающее элемент меню
        /// </returns>
        private static Func<ViewContextMenuContext, ValueTask> GetFilterMenuAction()
        {
            return c =>
                {
                    if (c.ViewContext.Parameters.Metadata.All(p => p.Hidden))
                    {
                        return new ValueTask();
                    }

                    c.MenuActions.Add(new MenuSeparatorAction(FilterSeparatorName));
                    c.MenuActions.Add(
                        new MenuAction(
                            FilterName,
                            LocalizationManager.Localize(UITilesFilter),
                            c.MenuContext.Icons.Get(FilterIconName),
                            new DelegateCommand(async o => await c.ViewContext.FilterViewAsync(), o => c.ViewContext.CanFilterView())));

                    return new ValueTask();
                };
        }

        /// <summary>
        ///     Возвращает действие создающее элемент меню
        /// </summary>
        /// <returns>
        ///     Действие создающее элемент меню
        /// </returns>
        private static Func<ViewContextMenuContext, ValueTask> GetRefreshMenuAction()
        {
            return
                c =>
                    {
                        c.MenuActions.Add(
                            new MenuAction(
                                RefreshName,
                                LocalizationManager.Localize(UITilesRefresh),
                                c.MenuContext.Icons.Get(RefreshIconName),
                                new DelegateCommand(
                                    async o => await c.ViewContext.RefreshViewAsync(),
                                    o => c.ViewContext.CanRefreshView())));

                        return new ValueTask();
                    };
        }

        /// <summary>
        ///     Возвращает действие создающее элемент меню
        /// </summary>
        /// <returns>
        ///     Действие создающее элемент меню
        /// </returns>
        private Func<ViewContextMenuContext, ValueTask> GetOpenCardMenuAction()
        {
            return async c =>
                {
                    if (c.ViewContext.RefSection is not null)
                    {
                        return;
                    }

                    var viewContext = c.ViewContext;
                    var view = viewContext.View;
                    IDictionary<string, object> selectedObject = viewContext.SelectedRow;
                    if (selectedObject == null || view == null)
                    {
                        return;
                    }

                    var metadata = await view.GetMetadataAsync(c.CancellationToken);
                    if (metadata == null)
                    {
                        return;
                    }

                    var cardRefs = (from r in metadata.References where r.IsCard select r).ToArray();
                    if (cardRefs.Length == 0)
                    {
                        return;
                    }

                    bool separatorAddided = false;
                    var addedIDs = new List<Guid>();
                    foreach (var cardRef in cardRefs)
                    {
                        var cardId = selectedObject.GetValueID(cardRef.ColPrefix);
                        if (cardId == null || cardId is DBNull)
                        {
                            return;
                        }

                        object displayValueObj = !string.IsNullOrWhiteSpace(cardRef.DisplayValueColumn)
                                                     ? selectedObject[cardRef.DisplayValueColumn.ToUpperInvariant()]
                                                     : selectedObject.GetFirstStringValueByPrefix(cardRef.ColPrefix);

                        string displayValue = displayValueObj != null ? displayValueObj.ToString() : string.Empty;

                        Guid id;
                        if (cardId is Guid guid)
                        {
                            id = guid;
                        }
                        else
                        {
                            if (!Guid.TryParse(cardId.ToString(), out id))
                            {
                                continue;
                            }
                        }

                        if (!addedIDs.Contains(id))
                        {
                            addedIDs.Add(id);

                            if (!separatorAddided)
                            {
                                c.MenuActions.Add(new MenuSeparatorAction(OpenCardSeparatorName));
                                separatorAddided = true;
                            }

                            c.MenuActions.Add(
                                new MenuAction(
                                    "OpenCard_" + id,
                                    string.Format("{0}: {1}", LocalizationManager.Localize(UITilesOpenCard), LocalizationManager.Format(displayValue)),
                                    c.MenuContext.Icons.Get(OpenCardIconName),
                                    new DelegateCommand(
                                        async o =>
                                        await c.MenuContext.UIContextExecutorAsync(async (ctx, ct) =>
                                            await this.OpenCardInUIHostAsync(id, displayValue, ctx, ct)))));
                        }
                    }
                };
        }

        /// <summary>
        /// Вызывает отображение карточки в UI
        /// </summary>
        /// <param name="id">
        /// Идентификатор карточки
        /// </param>
        /// <param name="displayValue">
        /// Отображаемая строка
        /// </param>
        /// <param name="context">
        /// Контекст
        /// </param>
        /// <param name="cancellationToken">Объект, посредством которого можно отменить асинхронную задачу.</param>
        /// <returns>Асинхронная задача.</returns>
        private async Task OpenCardInUIHostAsync(
            Guid id,
            string displayValue,
            IUIContext context,
            CancellationToken cancellationToken = default)
        {
            try
            {
                using ISplash splash = TessaSplash.Create(TessaSplashMessage.OpeningCard);

                IUIHost uiHost = this.hostGetter();
                await uiHost.OpenCardAsync(
                    id,
                    options: new OpenCardOptions
                    {
                        DisplayValue = displayValue,
                        UIContext = context,
                        Splash = splash,
                    },
                    cancellationToken: cancellationToken);
            }
            catch (NotSupportedException)
            {
                // используется FakeUIHost, например, мы открыты в TessaAdmin, игнорируем ошибку
            }
        }
    }
}